server/nbd-svc-routines.c

```diff
@@ -1124,44 +1124,63 @@ int nbd_handle_request(int sock, int threads)
     bzero(&nhdr, sizeof(struct nego_request));
     ret = nbd_socket_read(sock, &nhdr, sizeof(struct nego_request));
     if (ret != sizeof(struct nego_request)) {
+        nbd_err("Failed to read nego head!\n");
         ret = -1;
         goto err;
     }
 
     cfg = calloc(1, nhdr.len + 1);
+    if (!cfg) {
+        nbd_err("Failed to alloc memory for nhdr.len!\n");
+        ret = -1;
+        goto err;
+    }
     ret = nbd_socket_read(sock, cfg, nhdr.len);
     if (ret != nhdr.len) {
+        nbd_err("Failed to read nego head contents!\n");
         ret = -1;
-        free(cfg);
         goto err;
     }
 
     key = nbd_get_hash_key(cfg);
+
+    buf = calloc(1, 4096);
+    if (!buf) {
+        nbd_err("Failed to alloc memory for buf!\n");
+        ret = -1;
+        goto err;
+    }
     if (!key) {
         nrep.exit = -EINVAL;
-        buf = calloc(1, 4096);
         nrep.len = snprintf(buf, 4096, "Invalid cfg %s for nego!", cfg);
         if (nrep.len < 0)
             nrep.len = 0;
         nbd_err("Invalid cfg %s for nego!\n", cfg);
+        goto failed;
```

Comment:
Here we should just "goto err;", because if the key == NULL, we couldn't get the dev from the hash table, so it will make no sense to continue.

Reply:
But we need

1177     nbd_socket_write(sock, &nrep, sizeof(struct nego_reply));
1178     if (nrep.len)
1179         nbd_socket_write(sock, buf, nrep.len);

to submit nego_reply

Reply:
> Here we should just "goto err;", because if the key == NULL, we couldn't get the dev from the hash table, so it will make no sense to continue.

But we need

1177 nbd_socket_write(sock, &nrep, sizeof(struct nego_reply));
1178 if (nrep.len)
1179 nbd_socket_write(sock, buf, nrep.len);

to submit nego_reply

Reply:
Okay, I missed that part.